src/main/java/software/amazon/smithy/lsp/handler/CompletionHandler.java

```diff
@@ -0,0 +1,316 @@
+import software.amazon.smithy.lsp.project.Project;
+import software.amazon.smithy.lsp.project.SmithyFile;
+import software.amazon.smithy.lsp.protocol.RangeAdapter;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.loader.Prelude;
+import software.amazon.smithy.model.shapes.BlobShape;
+import software.amazon.smithy.model.shapes.BooleanShape;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.SetShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StringShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.TimestampShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.MixinTrait;
+import software.amazon.smithy.model.traits.RequiredTrait;
+import software.amazon.smithy.model.traits.TraitDefinition;
+
+public final class CompletionHandler {
+    // TODO: Handle keyword completions
+    private static final List<String> KEYWORDS = Arrays.asList("bigDecimal", "bigInteger", "blob", "boolean", "byte",
+            "create", "collectionOperations", "delete", "document", "double", "errors", "float", "identifiers", "input",
+            "integer", "integer", "key", "list", "long", "map", "member", "metadata", "namespace", "operation",
+            "operations",
+            "output", "put", "read", "rename", "resource", "resources", "service", "set", "short", "string",
+            "structure",
+            "timestamp", "union", "update", "use", "value", "version");
+
+    private final Project project;
+
+    public CompletionHandler(Project project) {
+        this.project = project;
+    }
+
+    /**
+     * @param params The request params
+     * @return A list of possible completions
+     */
+    public List<CompletionItem> handle(CompletionParams params, CancelChecker cc) {
+        String uri = params.getTextDocument().getUri();
+        SmithyFile smithyFile = project.getSmithyFile(uri);
+        if (smithyFile == null) {
+            return Collections.emptyList();
+        }
+
+        if (cc.isCanceled()) {
+            return Collections.emptyList();
+        }
+
+        Position position = params.getPosition();
+        CompletionContext completionContext = params.getContext();
+        if (completionContext != null
+            && completionContext.getTriggerKind().equals(CompletionTriggerKind.Invoked)
+            && position.getCharacter() > 0) {
+            // When the trigger is 'Invoked', the position is the next character
+            position.setCharacter(position.getCharacter() - 1);
+        }
+
+        if (cc.isCanceled()) {
```

Comment:
Yea I guess this is sort of an ooga way to do it. To answer your questions:
1. No, not necessarily. The longest running operations will be the last line, which actually computes the shapes -> completion items to return. Other than the underlying `Document` of a `SmithyFile`, everything else should be immutable, but honestly I just wrote this to check for cancellation whenever possible to early return. That seemed like the simplest way to do it.
2. These are all sync calls. I agree checking for cancellation everywhere is clunky, and I see that it isn't clear where these checks are absolutely necessary (making it more difficult to change later). We could do what you're saying, basically like a state machine, although I'm not sure how it would work when each 'step' adds some more state which needs to be passed into the next step. To me, that seems like maybe overkill for this, unless there was some other reason the current code is broken.

Reply:
TODO Add a todo to extract out some logic for a state-machine type model for this, where it checks for cancelation in between different states. That will be easier to understand and add to later, for someone who didn't write this code